### PR TITLE
Article - Using the Heap Data Structure in Python

### LINKED ISSUE #337: Article - Using the Binary Heap in Python ⚡
### Brief Summary: 
Describe the what, why, and how of your content idea in 2-5 sentences.

A binary heap is a special data structure that resembles a binary tree. It differs by the property that the root of any subtree should be the smallest or the largest element.
In a minheap, the root is the smallest element and in a maxheap, the root is the largest element.
In this article, we would learn how to implement a minheap and maxheap in Python.

### Key Takeaways: 
What are the 3-5 most important things the reader should understand or be able to do after reading this article?

- The reader would be able to understand the significance of the binary heap.
- The reader would understand the differences between a tree and a heap.
- The reader would be able to implement a binary heap in Python.

### References: 
Please list links to any published content/research that you intend to use to support/guide this article. (If none, please indicate 'N/A'.)
https://www.cpp.edu/~ftang/courses/CS241/notes/heap.htm

### PR DESCRIPTION
**Pull request by cyberShaw**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? see the example [Formats here](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)
- [x] Is your article properly formatted in Markdown? - see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)
- [x] Have you used the correct folder and file structure? - see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)
- [x] Is your article over 750 words?
- [x] Have you used our preferred heading structure? - [H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)
- [x] Does your article provide enough value and detail about your topic? Articles should be clear, accurate, and fully explained.
- [x] Can your article be understood by beginners? **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image? (600X400 pixels (images can be resized in paint) - also please make sure the image is from a site like www.unsplash.com)
- [x] In your PR - add a comment that states ["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278) to link and close your Idea Content (issue) when the PR is merged.
      

#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - Tool: [Grammarly](https://grammarly.com)
- [x] Have you checked your article for readability? - Tool: [Hemmingway](http://www.hemingwayapp.com/)
- [x] Have you added sources for quotes and images that aren't yours?

### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting/)
- [x] Have you used Repl.it for your code snippets?

### Contribution Guidelines
For first-time contributors and for more details, see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)

Added article on using the Heap Data Structure in Python. I've given a detailed description of Heaps and how to implement them using lists and the heapq library in Python.
This fixes #337

### CODE REVIEW

File: articles/heap-data-structure-python/index.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        print(self.heap) # Prints the heap
+```
+
+#### Minheap using Heapq
+
+We have successfully implemented a heap using a list. Now, let us use the **`heapq`** library in Python to implement a minheap.
+
+```python
+import heapq
+class Minheap:
```
Comment (botahamec) [inline articles/heap-data-structure-python/index.md:172]: For consistency, it's better to call this `MinHeap` with a capital H

File: articles/heap-data-structure-python/index.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    def __init__(self, minheap):
```
Comment (botahamec) [inline articles/heap-data-structure-python/index.md:173]: It's not clear what the reader should be passing into this constructor. What is `minheap`?

File: articles/heap-data-structure-python/index.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+  - This operation restored the heap property by rearranging the heap.
+  - Time complexity - O(log n).
+
+- **`printHeap()`**
+
+  - Prints the heap's level order traversal.
+
+#### Maxheap using List
+
+We are going to do the list implementation of a heap. In this, the heap's level-order traversal would be stored as an array/list.
```
Comment (botahamec) [inline articles/heap-data-structure-python/index.md:96]: The picture doesn't do a great job of explaining what Level-Order Traversal is. The Quora page you gave explains the order the heap is traversed in, which would help a lot.

**Review by botahamec — CHANGES_REQUESTED** [review]
Overall, this is a very good explanation of the heap. I only have a couple of suggestions for how to make your article a bit clearer.

File: articles/heap-data-structure-python/index.md (hunk 2 above)
Comment (cyberShaw) [reply articles/heap-data-structure-python/index.md:173]: I've added a code comment that specifies what minheap is.
It's a list that we pass as input. the list that we wanna convert to a heap.

File: articles/heap-data-structure-python/index.md (hunk 3 above)
Comment (cyberShaw) [reply articles/heap-data-structure-python/index.md:96]: > The picture doesn't do a great job of explaining what Level-Order Traversal is. The Quora page you gave explains the order the heap is traversed in, which would help a lot.

I've explained the diagram in detail and have included a link to the Quora page. 😄

**Comment by cyberShaw** [comment]
Thank you for the comprehensive review! 
I've made the requested changes in commit: 8c62445

**Review by botahamec — APPROVED** [review]
@hectorkambow This looks all good

**Comment by hectorkambow** [comment]
Good afternoon @gwritz  👋  Happy Friday - wanted to ask if you could give this "Using the Heap Data Structure in Python" article a final tech review - it has been peer-reviewed by one of our newest peer reviewers - but I wanted to get another set of eyes before we publish 🚀  
Thank you again!

**Review by gwritz — APPROVED** [review]

**Pull request merged**